frontends/concrete-python/concrete/fhe/mlir/context.py

```diff
@@ -271,42 +907,107 @@ def pack_to_chunk_groups_and_map(
             mapper (Callable):
                 mapping function
 
-            x_offset (int, default=0):
+            x_offset (int, default = 0):
                 optional offset for x during chunk extraction
 
-            y_offset (int, default=0):
+            y_offset (int, default = 0):
                 optional offset for x during chunk extraction
 
         Returns:
             List[Conversion]:
                 result of mapping chunks of x and y
         """
 
+        # this function computes the following in FHE
+        # -------------------------------------------
+        # result = []
+        # for chunk_index, (chunk_start, chunk_end) in enumerate(chunk_ranges):
+        #     x_chunk = (x + x_offset).bits[chunk_start:chunk_end]
+        #     y_chunk = (y + y_offset).bits[chunk_start:chunk_end]
+        #     result.append(mapper(chunk_index, x_chunk, y_chunk))
+        # return result
+        # -------------------------------------------
+        # - x_chunk :: { bit_width: (2 * chunk_size), shape: x.shape }
+        # - y_chunk :: { bit_width: (2 * chunk_size), shape: y.shape }
+        # - results :: { bit_width: resulting_scalar_type.bit_width, shape: resulting_shape }
+
+        assert resulting_scalar_type.is_scalar
+
         result = []
-        for chunk_index, offset in enumerate(range(0, bit_width, chunk_size)):
-            bits_to_process = min(chunk_size, bit_width - offset)
-            right_shift_by = bit_width - offset - bits_to_process
-            mask = (2**bits_to_process) - 1
+        for chunk_index, (chunk_start, chunk_end) in enumerate(chunk_ranges):
+            chunk_size = chunk_end - chunk_start
+
+            shift_to_clear_lsbs = chunk_start
+            mask_to_clear_msbs = (2**chunk_size) - 1
+
+            x_chunk_lut = [
+                (((x + x_offset) >> shift_to_clear_lsbs) & mask_to_clear_msbs) << chunk_size
+                for x in range(2**x.bit_width)
+            ]
+            y_chunk_lut = [
+                (((y + y_offset) >> shift_to_clear_lsbs) & mask_to_clear_msbs)
+                for y in range(2**y.bit_width)
+            ]
+
+            x_chunk_is_constant = all(x == x_chunk_lut[0] for x in x_chunk_lut)
+            y_chunk_is_constant = all(y == x_chunk_lut[0] for y in y_chunk_lut)
```

Comment:
y_chunk_lut[0] instead of x_chunk_lut[0]

Reply:
Bullet dodged I guess. Thanks!!!